amm/contracts/stable_pool/amp_coef.rs

```diff
@@ -0,0 +1,242 @@
+use amm_helpers::{
+    constants::stable_pool::{MAX_AMP, MAX_AMP_CHANGE, MIN_AMP, MIN_RAMP_DURATION},
+    ensure,
+};
+use ink::env::DefaultEnvironment;
+use traits::{MathError, StablePoolError};
+
+#[derive(Default, Debug, scale::Encode, scale::Decode, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "std",
+    derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout)
+)]
+pub struct AmpCoef {
+    /// Initial amplification coefficient.
+    init_amp_coef: u128,
+    /// Target for ramping up amplification coefficient.
+    future_amp_coef: u128,
+    /// Initial amplification time.
+    init_time: u64,
+    /// Stop ramp up amplification time.
+    future_time: u64,
+}
+
+impl AmpCoef {
+    pub fn new(init_amp_coef: u128) -> Result<Self, StablePoolError> {
+        ensure!(init_amp_coef >= MIN_AMP, StablePoolError::AmpCoefTooLow);
+        ensure!(init_amp_coef <= MAX_AMP, StablePoolError::AmpCoefTooHigh);
+        Ok(Self {
+            init_amp_coef,
+            future_amp_coef: init_amp_coef,
+            init_time: 0,
+            future_time: 0,
+        })
+    }
+
+    pub fn compute_amp_coef(&self) -> Result<u128, MathError> {
+        let current_time = ink::env::block_timestamp::<DefaultEnvironment>();
+        if current_time < self.future_time {
+            let time_range = self
+                .future_time
+                .checked_sub(self.init_time)
+                .ok_or(MathError::SubUnderflow(51))?;
+            let time_delta = current_time
+                .checked_sub(self.init_time)
+                .ok_or(MathError::SubUnderflow(52))?;
+
+            // Compute amp factor based on ramp time
+            let amp_range = self.future_amp_coef.abs_diff(self.init_amp_coef);
+            let amp_delta = amp_range
+                .checked_mul(time_delta as u128)
+                .ok_or(MathError::MulOverflow(51))?
+                .checked_div(time_range as u128)
+                .ok_or(MathError::DivByZero(51))?;
+            if self.future_amp_coef >= self.init_amp_coef {
+                // Ramp up
+                self.init_amp_coef
+                    .checked_add(amp_delta)
+                    .ok_or(MathError::AddOverflow(1))
+            } else {
+                // Ramp down
+                self.init_amp_coef
+                    .checked_sub(amp_delta)
+                    .ok_or(MathError::SubUnderflow(55))
+            }
+        } else {
+            Ok(self.future_amp_coef)
```

Comment:
Wouldn't it be OK to return `self.init_amp_coef` as well? We modify it in the `{ }` above, if we're still ramping up/down. It just seems weird to return "future" as the target.

Reply:
If there was ramping, there's a difference between the `init_amp_coef` and the `future_amp_coef`. Over the ramping period the `init_amp_coef +/- amp_delta` gets closer to the `future_amp_coef` as `current_time` increases. When the ramping is over it should return  the `future_amp_coef` because `future_time <= current_time` (the `else {...}` block).

Maybe we can reconsider the naming: rename `future_*` to `target_*`?

Reply:
Ah, I didn't notice we don't return from the first block. 

When the ramping is over, isn't it the case that `init_amp_coef` and `future_amp_coef` will be equal? 

Also, while the ramping lasts, this method will return _future_ value of Amp coef – is that correct? Imagine ramping is spread across 1000 blocks, so every block the A increases by 1/1000 (a `a_delta`) . We'd return the `A_init + 1000 * a_delta` here though.

Reply:
> When the ramping is over, isn't it the case that init_amp_coef and future_amp_coef will be equal?

Only if the ramping was stopped with `stop_ramp_amp_coef(...)` which sets both `init_amp_coef` and `future_amp_coef` to the value computed for the current time (block).

> Also, while the ramping lasts, this method will return future value of Amp coef – is that correct?

When it lasts (current_time < future_time) it will return `init_amp_coef +/- amp_delta`.
The value of `init_amp_coef` does not change after the ramping ends because is not necessary since we know that after the ramping ends the value of the `A` should be equal to the `future_amp_coef`

> Imagine ramping is spread across 1000 blocks, so every block the A increases by 1/1000 (a a_delta) . We'd return the A_init + 1000 * a_delta here though.

Something like that. Let's assume that A = 11, the current_block = 50 and we want to ramp it to 100 when we reach 1000 block. The values are set to:
- init_amp_coef = 11
- init_time = 50
- future_amp_coef = 100
- future_time = 1000

At a given block `N` where `N` < future_time, we can compute A like this:
A = init_amp_coef + (future_amp_coef - init_amp_coef) * (`N` - init_time) / (future_time - init_time) rounded down

E.g. `N` = 500:  11 + (100 - 11) * (500 - 50) / (1000 - 50) = ~53,16 = 53

If `N` is greater than future_time we should return the `future_amp_coef`.

Because we can't handle floats we can't tell by how much A will change every block (using integers) but we are able to tell what the value of A should be at a given block and that's how the value of A is computed in the ramping period.